go/runtime/scheduling/simple/txpool/priorityqueue/priority_queue.go

```diff
@@ -0,0 +1,218 @@
+// Package priorityqueue implements a tx pool backed by a priority queue.
+package priorityqueue
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/google/btree"
+
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+	"github.com/oasisprotocol/oasis-core/go/runtime/scheduling/simple/txpool/api"
+	"github.com/oasisprotocol/oasis-core/go/runtime/transaction"
+)
+
+// Name is the name of the tx pool implementation.
+const Name = "priority-queue"
+
+type item struct {
+	tx *transaction.CheckedTransaction
+}
+
+func (i item) Less(other btree.Item) bool {
+	i2 := other.(*item)
+	if p1, p2 := i.tx.Priority(), i2.tx.Priority(); p1 != p2 {
+		return p1 > p2
+	}
+	// If transactions have same priority, sort arbitrary.
+	h1 := i.tx.Hash()
+	h2 := i2.tx.Hash()
+	return bytes.Compare(h1[:], h2[:]) < 0
+}
+
+type priorityQueue struct {
+	sync.Mutex
+
+	priorityIndex *btree.BTree
+	transactions  map[hash.Hash]*item
+
+	maxTxPoolSize uint64
+
+	weightLimits map[string]uint64
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) Name() string {
+	return Name
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) Add(tx *transaction.CheckedTransaction) error {
+	q.Lock()
+	defer q.Unlock()
+
+	// Check if there is room in the queue.
+	if uint64(q.priorityIndex.Len()) >= q.maxTxPoolSize {
+		return api.ErrFull
+	}
+
+	if err := q.checkTxLocked(tx); err != nil {
+		return err
+	}
+
+	item := &item{tx: tx}
+	q.priorityIndex.ReplaceOrInsert(item)
+	q.transactions[tx.Hash()] = item
+
+	return nil
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) GetBatch(force bool) []*transaction.CheckedTransaction {
+	q.Lock()
+	defer q.Unlock()
+
+	// Check if a batch is ready.
+	// XXX: could also check if any other weight limits are reached.
```

Comment:
Yeah this probably makes sense, but that would require keeping per-weight tallies for what is in the current pool, right? But that would generalize this nicely and wouldn't hardcode a `.Len()` here.

Reply:
Indeed, fixed.